### PR TITLE
Fix route replacement when service is missing

### PR DESCRIPTION
**Pull request by jenshu**
# Description

During gloo translation, a [snapshot sanitizer](https://github.com/solo-io/gloo/blob/f4255209a99293f386322664490ffdba194e4444/projects/gloo/pkg/syncer/setup/setup_syncer.go#L558) runs, which first runs an [upstream removing sanitizer](https://github.com/solo-io/gloo/blob/f4255209a99293f386322664490ffdba194e4444/projects/gloo/pkg/syncer/sanitizer/upstream_removing_sanitizer.go) and then a [route replacing sanitizer](https://github.com/solo-io/gloo/blob/f4255209a99293f386322664490ffdba194e4444/projects/gloo/pkg/syncer/sanitizer/route_replacing_sanitizer.go).

When we have an upstream that points to a missing service, the upstream removing sanitizer removes the upstream/cluster from the xds snapshot. Then the route replacing sanitizer is run, which creates a fallback cluster for any invalid routes. The bug was that the route replacing sanitizer was assuming all upstreams in the gloo snapshot were valid clusters. The fix is to *also* check that the cluster still exists in the xds snapshot (i.e. hadn't already been removed by another sanitizer). When the route replacing sanitizer sees that the cluster doesn't exist in the xds snapshot, it will create a fallback route.

# Context

Users ran into this bug doing ... \ Users needed this feature to ...

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gateway/pkg/translator/http_translator.go [hunk 1]
```diff
@@ -471,7 +471,7 @@ func regexShortCircuits(laterMatcher, earlierMatcher *matchers.Matcher) bool {
 	foundIndex := re.FindStringIndex(laterPath)
 	// later matcher is always non-regex. to validate against the regex, we need to ensure that it's either
 	// unset or set to false
-	return foundIndex != nil && (laterMatcher.GetCaseSensitive() == nil || laterMatcher.GetCaseSensitive().GetValue() == false)
+	return foundIndex != nil && laterMatcher.GetCaseSensitive().GetValue() == false
```
Comment (jenshu) [inline projects/gateway/pkg/translator/http_translator.go:474]: this is just addressing a nit from a previous (unrelated) PR

File: projects/gloo/pkg/syncer/sanitizer/upstream_removing_sanitizer.go [hunk 2]
```diff
@@ -54,6 +54,10 @@ func (s *UpstreamRemovingSanitizer) SanitizeSnapshot(
 		if reports[up].Errors != nil {
 
 			clusterName := translator.UpstreamToClusterName(up.GetMetadata().Ref())
+			if clusters.Items[clusterName] == nil {
```
Comment (jenshu) [inline projects/gloo/pkg/syncer/sanitizer/upstream_removing_sanitizer.go:57]: this change is unrelated to the route replacement bug. not sure exactly how to reproduce, but we have seen cases where the `clusters.Items[clusterName].ResourceProto()...` line below panics. just adding a check here to avoid that situation

**Review by nfuden — APPROVED** [review]
Thanks for the PR description! It not only made the rationale super clear it also helped answer some questions I almost put on the PR!
🥇
